File: 2023/masterkeygeneration.m
function [masterkeyA,masterkeyB]=masterkeygeneration(data,anchors,anchor1data,anchor2data,mastersetA,mastersetB,setwidth)
    
    %find length of anchors
    anchorsize=size(anchors);
    anchorlength=anchorsize(1);
    
    %Initialize arrays
    masterkeyA=zeros(12,12);
    masterkeyB=zeros(12,12);
    k1=0;
    k2=0;
    countsA=zeros(12,1);
    countsB=zeros(12,1);
    for i = 1:anchorlength
        for j=1:12
            if setwidth(i,1)==j
                if anchors(i, 2) == 1
                    k1=k1+1;
                    countsA(j,1)=countsA(j,1)+1;
                    for m = 1:setwidth(i,1)
                        masterkeyA(j, m) = masterkeyA(j, m) + mastersetA(k1, m);
                    end
                elseif anchors(i,2) == 2
                    k2=k2+1;
                    countsB(j,1)=countsB(j,1)+1;
                    for m = 1:setwidth(i,1)
                        masterkeyB(j, m) = masterkeyB(j, m) + mastersetB(k2, m);
                    end
                end
            end
        end
    end
    
    %create masterkey
    for i = 1:12
        for j=1:12
            masterkeyA(i, j) = masterkeyA(i, j) / countsA(i,1);
            masterkeyB(i, j) = masterkeyB(i, j) / countsB(i,1);
        end
    end
end

File: 2024/masterkeygeneration.py
import numpy as np

def masterkeygeneration(data, anchors, mastersetA, mastersetB, setwidth):
  """
  Generates two master keys based on trend differences and widths.

  Args:
    data: A numpy array containing the data points.
    anchors: A numpy array with anchor points, each row containing (index, flag).
      - index: Index of the data point in the data array.
      - flag: 1 for increasing trend, 2 for decreasing trend.
    mastersetA: A numpy array containing trend differences for increasing trends.
    mastersetB: A numpy array containing trend differences for decreasing trends.
    setwidth: A list containing the number of data points in each trend.

  Returns:
    masterkeyA: A numpy array representing the master key for increasing trends.
    masterkeyB: A numpy array representing the master key for decreasing trends.
  """

  anchor_size = len(anchors)

  # Initialize arrays
  masterkeyA = np.zeros((12, 12))
  masterkeyB = np.zeros((12, 12))
  countsA = np.zeros(12)
  countsB = np.zeros(12)

  k1 = 0
  k2 = 0

  for i in range(anchor_size):
    width = setwidth[i]
    if width <= 12:
      if anchors[i][1] == 1:
        k1 += 1
        countsA[width - 1] += 1
        masterkeyA[width - 1, :width] += mastersetA[k1 - 1, :width]
      elif anchors[i][1] == 2:
        k2 += 1
        countsB[width - 1] += 1
        masterkeyB[width - 1, :width] += mastersetB[k2 - 1, :width]

  # Create master key by averaging
  for i in range(12):
    for j in range(12):
      if countsA[i] > 0:
        masterkeyA[i, j] /= countsA[i]
      if countsB[i] > 0:
        masterkeyB[i, j] /= countsB[i]

  return masterkeyA, masterkeyB

